upstairs/src/lib.rs

```diff
@@ -4475,26 +4487,23 @@ impl Downstairs {
             assert!(!self.completed.contains(&ds_id));
             assert_eq!(wc.active, 0);
 
-            // Sort the job list, and retire all the jobs that happened before
-            // and including this flush.
-
-            let mut kvec: Vec<u64> = self
-                .ds_active
-                .keys()
-                .cloned()
-                .filter(|&x| x <= ds_id)
-                .collect::<Vec<u64>>();
+            // Retire all the jobs that happened before and including this
+            // flush.
+            let mut retired = Vec::new();
 
-            kvec.sort_unstable();
+            loop {
+                let id = match self.ds_active.keys().next() {
```

Comment:
I guess this will still be sorted by key?  I'm not sure how we are making sure that
the first key we get that is above our flush allows us to break unless we are sure
here that `self.ds_active.keys().next()` will iterate from lowest to highest key value.

It may just be that, by the nature of our insertion/creation of keys that they always
end up in order, but if so, perhaps we need to comment that somewhere so we
don't accidentally lose that order that we now depend on here.